From f3b1ce2f372d6d118df9f7fdaaae7e8db7759af9 Mon Sep 17 00:00:00 2001
From: Vinit Agnihotri <vagnihot@redhat.com>
Date: Mon, 23 Mar 2026 21:01:37 +0530
Subject: [PATCH] printing: Set default value in case of non-exisiting record

This fixes regression caused by commit#e9a7dce599
Newly added function treats non-existing record as error, instead of
setting just -1 i.e. default value for non-exisiting record for
printing subsystem. This results into print_cache_expired returning
incorrect status.

Fix sets default value in case of non-exisiting record
which would ensure print_cache_expired to return correct status.

BUG: https://bugzilla.samba.org/show_bug.cgi?id=16038

Signed-off-by: Vinit Agnihotri <vagnihot@redhat.com>
Reviewed-by: Rabinarayan Panigrahi <rapanigr@redhat.com>
Reviewed-by: Avan Thakkar <athakkar@redhat.com>
Reviewed-by: Xavi Hernandez <xhernandez@redhat.com>
Reviewed-by: Anoop C S <anoopcs@samba.org>
Reviewed-by: Guenther Deschner <gd@samba.org>
---
 source3/printing/printing.c | 21 +++++++++++++--------
 1 file changed, 13 insertions(+), 8 deletions(-)

diff --git a/source3/printing/printing.c b/source3/printing/printing.c
index d92ebd04a55..5161aa54c31 100644
--- a/source3/printing/printing.c
+++ b/source3/printing/printing.c
@@ -61,22 +61,28 @@ static int fetch_share_cache_time(const char *key_name,
 {
 	char *key = NULL;
 	int64_t curr_time64 = -1;
+	int ret = -1;
 
 	key = talloc_asprintf(NULL, "%s/%s", key_name, sharename);
 	if (key == NULL) {
 		DBG_ERR("Failed to format key\n");
-		return -1;
+		goto out;
 	}
 
-	if (tdb_fetch_int64(tdb, key, &curr_time64) != 0) {
-		DBG_ERR("No timing record found for[%s]!\n", sharename);
-		TALLOC_FREE(key);
-		return -1;
+	ret = tdb_fetch_int64(tdb, key, &curr_time64);
+	if (ret != 0) {
+		enum TDB_ERROR tdb_err = tdb_error(tdb);
+		if (tdb_err != TDB_ERR_NOEXIST) {
+			DBG_ERR("Error fetching timing data for [%s]. Err:%s\n",
+				sharename, tdb_errorstr(tdb));
+			goto out;
+		}
 	}
 	*curr_time = curr_time64;
-
+	ret = 0;
+out:
 	TALLOC_FREE(key);
-	return 0;
+	return ret;
 }
 
 static int update_share_cache_time(const char *key_name,
@@ -1312,7 +1318,6 @@ static bool print_cache_expired(const char *sharename, bool check_pending)
 
 		msg_pending_time = u;
 		if ( check_pending
-			&& u
 			&& msg_pending_time > 0
 			&& msg_pending_time <= time_now
 			&& (time_now - msg_pending_time) < 60 )
-- 
2.53.0

